### PR TITLE
rewrite charm to sidecar

### PR DESCRIPTION
**Pull request by beliaev-maksim**
rewrite OIDC charm to sidecar instead of podspec

Note, there was fundamental issue (or indented feature?) that is changed now.

Before, `_check_public_url` was not called at all anywhere in the code. Which allowed execution without this config and made config optional

This rewrite enforces the config to be applied, otherwise unit will be in Blocked state

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -7,8 +7,11 @@ description: |
   asks if an incoming request is authorized.
 
   https://journal.arrikto.com/kubeflow-authentication-with-istio-dex-5eafdfac4782
-series: [kubernetes]
-min-juju-version: "2.9.0"
+
```
Comment (i-chvets) [inline metadata.yaml:10]: Add doc links.
```suggestion
website: https://charmhub.io/oidc-gatekeeper
source: https://github.com/canonical/oidc-gatekeeper-operator
issues: https://github.com/canonical/oidc-gatekeeper-operator/issues
```

File: src/charm.py [hunk 2]
```diff
@@ -1,24 +1,41 @@
...
+from lightkube.models.core_v1 import ServicePort
 from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer
 from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
 
 
 class Operator(CharmBase):
```
Comment (i-chvets) [inline src/charm.py:20]: Should this be `OidcGatekeeperOperator` instead of `Operator`?

File: src/charm.py [hunk 3]
```diff
@@ -1,24 +1,41 @@
...
     def __init__(self, *args):
```
Comment (i-chvets) [inline src/charm.py:21]: Doc string is missing.

File: src/charm.py (hunk 2 above)
Comment (i-chvets) [inline src/charm.py:20]: Doc string is missing.

File: src/charm.py [hunk 4]
```diff
@@ -179,23 +180,12 @@ def _check_secret(self, event=None):
...
-    def __init__(self, msg, status_type=None):
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
-
-
 if __name__ == "__main__":
     main(Operator)
```
Comment (i-chvets) [inline src/charm.py:191]: Change `Opeator` to `OidcGatekeeperOperator`.

File: src/charm.py [hunk 5]
```diff
@@ -34,105 +51,89 @@ def __init__(self, *args):
...
-        except OCIImageResourceError as e:
-            raise CheckFailed(f"{e.status_message}: oci-image", e.status_type)
-        return image_details
-
     def _check_public_url(self):
         if not self.model.config.get("public-url"):
-            raise CheckFailed("public-url config required", BlockedStatus)
+            raise ErrorWithStatus("public-url config required", BlockedStatus)
 
     def _configure_mesh(self, interfaces):
```
Comment (i-chvets) [inline src/charm.py:138]: Doc string is missing.

File: src/charm.py [hunk 6]
```diff
@@ -179,23 +180,12 @@ def _check_secret(self, event=None):
                 rel.data[self.model.app]["client-secret"] = _gen_pass()
             return rel.data[self.model.app]["client-secret"]
         else:
-            raise CheckFailed("Waiting for Client Secret", WaitingStatus)
+            raise ErrorWithStatus("Waiting for Client Secret", WaitingStatus)
 
 
 def _gen_pass() -> str:
```
Comment (i-chvets) [inline src/charm.py:186]: Doc string is missing.

File: src/charm.py [hunk 7]
```diff
@@ -1,24 +1,41 @@
...
+            self,
+            [http_port],
+        )
+
+        self.public_url = self.model.config["public-url"]
+        if not self.public_url.startswith(("http://", "https://")):
+            self.public_url = f"http://{self.public_url}"
 
         for event in [
             self.on.start,
```
Comment (i-chvets) [inline src/charm.py:41]: Can we handle upgrade with this setup or do we need separate upgrade event handler?

File: tests/integration/test_charm.py [hunk 8]
```diff
@@ -44,22 +48,10 @@ async def test_relations(ops_test: OpsTest):
...
 
-    await ops_test.model.wait_for_idle(
-        [APP_NAME, ISTIO_PILOT, DEX_AUTH],
-        status="active",
-        raise_on_blocked=True,
-        raise_on_error=True,
-        timeout=600,
-    )
-
-
```
Comment (i-chvets) [inline tests/integration/test_charm.py:55]: Add upgrade test.

File: tests/integration/test_charm.py [hunk 9]
```diff
@@ -44,22 +48,10 @@ async def test_relations(ops_test: OpsTest):
...
     await ops_test.model.add_relation(f"{ISTIO_PILOT}:ingress-auth", f"{APP_NAME}:ingress-auth")
 
-    await ops_test.model.wait_for_idle(
-        [APP_NAME, ISTIO_PILOT, DEX_AUTH],
-        status="active",
-        raise_on_blocked=True,
-        raise_on_error=True,
-        timeout=600,
-    )
-
```
Comment (i-chvets) [inline tests/integration/test_charm.py:54]: Istio pilot should be `1.16/stable`

Comment (beliaev-maksim) [reply tests/integration/test_charm.py:54]: let's see if that will work. I did not touch tests as they were

Comment (i-chvets) [inline tests/integration/test_charm.py:54]: Should relation between Dex and Oidc Gatekeeper be also tested?
`await ops_test.model.add_relation(f"{APP_NAME}:oidc-client", f"{DEX_AUTH}:oidc-client")`

Comment (beliaev-maksim) [reply tests/integration/test_charm.py:54]: cannot bump due  to the bug https://github.com/juju/python-libjuju/issues/851

Comment (i-chvets) [reply tests/integration/test_charm.py:54]: Leaving as is then.

**Comment by i-chvets** [comment]
Integration tests pass locally:
```
$ tox -e integration
integration: install_deps> python -I -m pip install -r requirements-integration.txt
integration: commands[0]> pytest -v --tb native --asyncio-mode=auto /home/ichvets/cw/dev/oidc-gatekeeper-operator/tests/integration --log-cli-level=INFO -s
======================================================== test session starts =========================================================
platform linux -- Python 3.10.6, pytest-7.2.0, pluggy-1.0.0 -- /home/ichvets/cw/dev/oidc-gatekeeper-operator/.tox/integration/bin/python
cachedir: .tox/integration/.pytest_cache
rootdir: /home/ichvets/cw/dev/oidc-gatekeeper-operator, configfile: pyproject.toml
plugins: operator-0.22.0, anyio-3.6.2, asyncio-0.20.3
asyncio: mode=auto
collecting ... collected 4 items

tests/integration/test_charm.py::TestOIDCOperator::test_build 
----------------------------------------------------------- live log setup -----------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:625 Adding model uk8s:test-charm-19xa on cloud microk8s
----------------------------------------------------------- live log call ------------------------------------------------------------
INFO     pytest_operator.plugin:plugin.py:504 Using tmp_path: /home/ichvets/cw/dev/oidc-gatekeeper-operator/.tox/integration/tmp/pytest/test-charm-19xa0
INFO     pytest_operator.plugin:plugin.py:948 Building charm oidc-gatekeeper
INFO     pytest_operator.plugin:plugin.py:953 Built charm oidc-gatekeeper in 182.00s
PASSED
tests/integration/test_charm.py::TestOIDCOperator::test_deploy 
----------------------------------------------------------- live log call ------------------------------------------------------------
INFO     juju.model:model.py:1959 Deploying local:oidc-gatekeeper-0
INFO     juju.model:model.py:2582 Waiting for model:
  oidc-gatekeeper/0 [allocating] waiting: installing agent
INFO     juju.model:model.py:2582 Waiting for model:
  oidc-gatekeeper/0 [allocating] waiting: agent initializing
INFO     juju.model:model.py:2582 Waiting for model:
  oidc-gatekeeper/0 [idle] active: 
PASSED
tests/integration/test_charm.py::TestOIDCOperator::test_relations 
----------------------------------------------------------- live log call ------------------------------------------------------------
INFO     juju.model:model.py:1959 Deploying ch:amd64/focal/istio-pilot-63
INFO     juju.model:model.py:1959 Deploying ch:amd64/focal/dex-auth-224
INFO     juju.model:model.py:2582 Waiting for model:
  oidc-gatekeeper/0 [executing] active: 
  istio-pilot/0 [allocating] waiting: installing agent
  dex-auth/0 [allocating] waiting: installing agent
INFO     juju.model:model.py:2582 Waiting for model:
  istio-pilot/0 [allocating] waiting: installing agent
  dex-auth/0 [allocating] waiting: agent initializing
INFO     juju.model:model.py:2582 Waiting for model:
  istio-pilot/0 [allocating] waiting: agent initializing
  dex-auth/0 [executing] maintenance: installing charm software
INFO     juju.model:model.py:2582 Waiting for model:
  oidc-gatekeeper/0 [executing] waiting: List of oidc-client versions not found for apps: dex-auth
  istio-pilot/0 [executing] active: 
  dex-auth/0 [executing] maintenance: installing charm software
INFO     juju.model:model.py:2582 Waiting for model:
  oidc-gatekeeper/0 [idle] active: 
  istio-pilot/0 [idle] active: 
  dex-auth/0 [idle] active: 
PASSED
```

**Review by i-chvets — DISMISSED** [review]
Approved.
ACK: CI is failing in Github runner, passes locally.

**Review by i-chvets — APPROVED** [review]
Approved.

**Pull request merged**